README.md
sparklyr.sedona
================
[![Github workflow status](https://github.com/r-spark/sparklyr.sedona/actions/workflows/ci.yaml/badge.svg)](https://github.com/r-spark/sparklyr.sedona/actions/workflows/ci.yaml)

## Overview

Sparklyr.sedona is a [sparklyr](https://github.com/sparklyr/sparklyr)-based R interface for [Apache Sedona](https://sedona.apache.org).
It presents what Apache Sedona has to offer through idiomatic frameworks and constructs in R
(e.g., one can build spatial Spark SQL queries using Sedona UDFs in conjunction with a wide range of `dplyr` expressions),
hence making Apache Sedona highly friendly for R users.

## Connecting to Spark

To ensure Sedona serialization routines, UDTs, and UDFs are properly registered when creating a Spark session, one simply needs to attach `sparklyr.sedona` before
instantiating a Spark conneciton. Sparklyr.sedona will take care of the rest. For example,

``` r
library(sparklyr)
library(sparklyr.sedona)

spark_home <- "/usr/lib/spark"  # NOTE: replace this with your $SPARK_HOME directory
sc <- spark_connect(master = "yarn", spark_home = spark_home)
```

will create a Sedona-capable Spark connection in YARN client mode, and

``` r
library(sparklyr)
library(sparklyr.sedona)

sc <- spark_connect(master = "local")
```

will create a Sedona-capable Spark connection to an Apache Spark instance running locally.

In `sparklyr`, one can easily inspect the Spark connection object to sanity-check it has been properly initialized with all Sedona-related dependencies, e.g.,

``` r
print(sc$extensions$packages)
```

```
## [1] "org.apache.sedona:sedona-core-3.0_2.12:1.0.0-incubating"
## [2] "org.apache.sedona:sedona-sql-3.0_2.12:1.0.0-incubating"
## [3] "org.apache.sedona:sedona-viz-3.0_2.12:1.0.0-incubating"
## [4] "org.datasyslab:geotools-wrapper:geotools-24.0"
## [5] "org.datasyslab:sernetcdf:0.1.0"
## [6] "org.locationtech.jts:jts-core:1.18.0"
## [7] "org.wololo:jts2geojson:0.14.3"
```

and

``` r
spark_session(sc) %>%
  invoke("%>%", list("conf"), list("get", "spark.kryo.registrator")) %>%
  print()
```

```
## [1] "org.apache.sedona.viz.core.Serde.SedonaVizKryoRegistrator"
```
.

For more information about connecting to Spark with `sparklyr`, see https://therinspark.com/connections.html and `?sparklyr::spark_connect`.
Also see https://sedona.apache.org/tutorial/rdd/#initiate-sparkcontext for minimum and recommended dependencies for Apache Sedona.

## What are `SpatialRDD`s?

[SpatialRDD](https://sedona.apache.org/api/javadoc/core/org/apache/sedona/core/spatialRDD/SpatialRDD.html)s are basic building blocks of
distributed spatial data in Apache Sedona. A `SpatialRDD` can be spatially partitioned and indexed to facilitate range queries, KNN queries,
and other low-level operations. One can also export records from `SpatailRDD`s into regular Spark dataframes, making them accessible through
Spark SQL and through the `dplyr` interface of `sparklyr` (more on that later).

## Creating a SpatialRDD

NOTE: this section is largely based on https://sedona.apache.org/tutorial/rdd/#create-a-spatialrdd, except for examples have been
written in R instead of Scala to reflect usages of `sparklyr.sedona`.

Currently `SpatialRDD`s can be created in `sparklyr.sedona` by reading a file in a supported geospatial format, or by extracting data from a
Spark SQL query.

For example, the following code will import data from [arealm-small.csv](https://github.com/apache/incubator-sedona/blob/master/binder/data/arealm-small.csv) into a `SpatialRDD`:

``` r
pt_rdd <- sedona_read_dsv_to_typed_rdd(
  sc,
  location = "arealm-small.csv",
  delimiter = ",",
  type = "point",
  first_spatial_col_index = 1,
  has_non_spatial_attrs = TRUE
)

```
.

Records from the example [arealm-small.csv](https://github.com/apache/incubator-sedona/blob/master/binder/data/arealm-small.csv) file look like the following:

```
testattribute0,-88.331492,32.324142,testattribute1,testattribute2
testattribute0,-88.175933,32.360763,testattribute1,testattribute2
testattribute0,-88.388954,32.357073,testattribute1,testattribute2
```

As one can see from the above, each record is comma-separated and consists of a 2-dimensional coordinate starting at the 2nd column and ending at the 3rd column.
All other columns contain non-spatial attributes. Because column indexes are 0-based, we need to specify `first_spatial_col_index = 1` in the example above to
ensure each record is parsed correctly.

In addition to formats such as CSV and TSV, currently `sparklyr.sedona` also supports reading files in WKT (Well-Known Text), WKB (Well-Known Binary), and GeoJSON formats.
See `?sparklyr.sedona::sedona_read_wkt`, `?sparklyr.sedona::sedona_read_wkb`, and `?sparklyr.sedona::sedona_read_geojson` for details.

One can also run `to_spatial_rdd()` to extract a SpatailRDD from a Spark SQL query, e.g.,

``` r
library(sparklyr)
library(sparklyr.sedona)
library(dplyr)

sc <- spark_connect(master = "local")

sdf <- tbl(
  sc,
  sql("SELECT ST_GeomFromText('POINT(-71.064544 42.28787)') AS `geom`, \"point\" AS `type`")
)

spatial_rdd <- sdf %>% to_spatial_rdd(spatial_col = "geom")
print(spatial_rdd)
```

```
## $.jobj
## <jobj[70]>
##   org.apache.sedona.core.spatialRDD.SpatialRDD
##   org.apache.sedona.core.spatialRDD.SpatialRDD@422afc5a
##
## ...
```

will extract a spatial column named `"geom"` from the Sedona spatial SQL query above and store it in a `SpatialRDD` object.

## Working with Spark dataframes

As mentioned previously, data from `SpatialRDD` can be exported into a Spark dataframe and be queried and modified through
the `dplyr` interface of `sparklyr`. The example below shows how `sdf_register()`, a S3 generic that converts a lower-level
object into a Spark dataframe object in `sparklyr`, can be applied to a `SpatialRDD` object created by `sparklyr.sedona`.

``` r
library(sparklyr)
library(sparklyr.sedona)

sc <- spark_connect(master = "local")
polygon_rdd <- sedona_read_geojson(sc, location = "/tmp/polygon.json")
polygon_sdf <- polygon_rdd %>% sdf_register()

polygon_sdf %>% print(n = 3)
```

```
## # Source: spark<?> [?? x 1]
##   geometry
##   <list>
## 1 <POLYGON ((-87.621765 34.873444, -87.617535 34.873369, -87.6123 34.873337, -8…
## 2 <POLYGON ((-85.719017 31.297901, -85.715626 31.305203, -85.714271 31.307096, …
## 3 <POLYGON ((-86.000685 34.00537, -85.998837 34.009768, -85.998012 34.010398, -…
## # … with more rows
```

The Spark dataframe object can then be modified using `dplyr` verbs familiar to many R users. In addition, spatial UDFs
supported by Sedona can inter-operate seamlessly with other functions supported in `sparklyr`'s dbplyr SQL translation
env. For example, the code below finds the average area of all polygons in `polygon_sdf`:

``` r
mean_area_sdf <- polygon_sdf %>%
  dplyr::summarize(dplyr::summarize(mean_area = mean(ST_Area(geometry))))
print(mean_area_sdf)
```

```
## # Source: spark<?> [?? x 1]
##   mean_area
##       <dbl>
## 1   0.00217
```

Once spatial objects are imported into Spark dataframes, they can also be easily integrated with other non-spatial attributes,
e.g.,

``` r
modified_polygon_sdf <- polygon_sdf %>%
  dplyr::mutate(type = "polygon")
```
.

Notice all of the above can open up many interesting possiblities. For example, one can extract ML features from geospatial
data in Spark dataframes, build a ML pipeline using `ml_*` family of functions in `sparklyr` to work with such features, and if
the output of a ML model happens to be a geospatial object as well, one can even apply visualization routines in
`sparklyr.sedona` to visualize the difference between any predicted geometry and the corresponding ground truth
(more on visualization later).

## Visualization

It is worth mentioning an important part of `sparklyr.sedona` is its collection of R interfaces to Sedona visualization routines.
For example, the following
is essentially the R equivalent of [this example in Scala](https://github.com/apache/incubator-sedona/blob/f6b1c5e24bdb67d2c8d701a9b2af1fb5658fdc4d/viz/src/main/scala/org/apache/sedona/viz/showcase/ScalaExample.scala#L142-L160).

``` r
library(sparklyr)
library(sparklyr.sedona)

sc <- spark_connect(master = "local")

resolution_x <- 1000
resolution_y <- 600
boundary <- c(-126.790180, -64.630926, 24.863836, 50.000)

pt_rdd <- sedona_read_dsv_to_typed_rdd(
  sc,
  location = "arealm.csv",
  type = "point"
)
polygon_rdd <- sedona_read_dsv_to_typed_rdd(
  sc,
  location = "primaryroads-polygon.csv",
  type = "polygon"
)
pair_rdd <- sedona_spatial_join_count_by_key(
  pt_rdd,
  polygon_rdd,
  join_type = "intersect"
)

overlay <- sedona_render_scatter_plot(
  polygon_rdd,
  resolution_x,
  resolution_y,
  output_location = tempfile("scatter-plot-"),
  boundary = boundary,
  base_color = c(255, 0, 0),
  browse = FALSE
)

sedona_render_choropleth_map(
  pair_rdd,
  resolution_x,
  resolution_y,
  output_location = "/tmp/choropleth-map",
  boundary = boundary,
  overlay = overlay,
  # vary the green color channel according to relative magnitudes of data points so
  # that the resulting map will show light blue, light purple, and light gray pixels
  color_of_variation = "green",
  base_color = c(225, 225, 255)
)
```

It will create a scatter plot, and then overlay it on top of a choropleth map, as shown below:

<img src="docs/choropleth-map.png" width=800 />

See `?sparklyr.sedona::sedona_render_scatter_plot`, `?sparklyr.sedona::sedona_render_heatmap`,
and `?sparklyr.sedona::sedona_render_choropleth_map` for more details on R interfaces of
Sedona visualization routines currently implemented by `sparklyr.sedona`.


NEWS.md
# sparklyr.sedona

`sparklyr.sedona` is a `sparklyr` extension that provides an R interface fo
 [Apache Sedona](https://sedona.apache.org/).

In May 2021, `sparklyr.sedona` has been migrated to
[here](https://github.com/apache/incubator-sedona/tree/master/R/sparklyr.sedona) to officially
become part of the [https://github.com/apache/incubator-sedona](incubator-sedona) project.


R/viz.R
#' Visualization routine for Sedona spatial RDD.
#'
#' Generate a visual representation of geometrical object(s) within a Sedona
#' spatial RDD.
#'
#' @param rdd A Sedona spatial RDD.
#' @param resolution_x Resolution on the x-axis.
#' @param resolution_y Resolution on the y-axis.
#' @param output_location Location of the output image. This should be the
#'   desired path of the image file excluding extension in its file name.
#' @param output_format File format of the output image. Currently "png",
#'   "gif", and "svg" formats are supported (default: "png").
#' @param boundary Only render data within the given rectangular boundary.
#'   The `boundary` parameter can be set to either a numeric vector of
#'   c(min_x, max_y, min_y, max_y) values, or with a bounding box object
#'   e.g., new_bounding_box(sc, min_x, max_y, min_y, max_y), or NULL
#'   (the default). If `boundary` is NULL, then the minimum bounding box of the
#'   input spatial RDD will be computed and used as boundary for rendering.
#' @param color_of_variation Which color channel will vary depending on values
#'   of data points. Must be one of "red", "green", or "blue". Default: red.
#' @param base_color Color of any data point with value 0. Must be a numeric
#'   vector of length 3 specifying values for red, green, and blue channels.
#'   Default: c(0, 0, 0).
#' @param shade Whether data point with larger magnitude will be displayed with
#'   darker color. Default: TRUE.
#' @param overlay A \code{viz_op} object containing a raster image to be
#'   displayed on top of the resulting image.
#' @param browse Whether to open the rendered image in a browser (default:
#'   interactive()).
#'
#' @name sedona_visualization_routines
NULL

#' Visualize a Sedona spatial RDD using a heatmap.
#'
#' Generate a heatmap of geometrical object(s) within a Sedona spatial RDD.
#'
#' @inheritParams sedona_visualization_routines
#' @param blur_radius Controls the radius of a Gaussian blur in the resulting
#'   heatmap.
#'
#' @examples
#'
#' library(sparklyr)
#' library(sparklyr.sedona)
#'
#' sc <- spark_connect(master = "spark://HOST:PORT")
#'
#' if (!inherits(sc, "test_connection")) {
#'   input_location <- system.file(
#'     file.path("extdata", "arealm-small.csv"), package = "sparklyr.sedona"
#'   )
#'   rdd <- sedona_read_dsv_to_typed_rdd(
#'     sc,
#'     location = input_location,
#'     type = "point"
#'   )
#'
#'   sedona_render_heatmap(
#'     rdd,
#'     resolution_x = 800,
#'     resolution_y = 600,
#'     output_location = tempfile("arealm-small-"),
#'     output_format = "png",
#'     boundary = c(-91, -84, 30, 35),
#'     blur_radius = 10
#'   )
#' }
#'
#' @family Sedona visualization routines
#' @export
sedona_render_heatmap <- function(
                                  rdd,
                                  resolution_x,
                                  resolution_y,
                                  output_location,
                                  output_format = c("png", "gif", "svg"),
                                  boundary = NULL,
                                  blur_radius = 10L,
                                  overlay = NULL,
                                  browse = interactive()) {
  sc <- spark_connection(rdd$.jobj)
  output_format <- match.arg(output_format)

  boundary <- validate_boundary(rdd, boundary)
  viz_op <- invoke_new(
    sc,
    "org.apache.sedona.viz.extension.visualizationEffect.HeatMap",
    as.integer(resolution_x),
    as.integer(resolution_y),
    boundary$.jobj,
    FALSE,
    as.integer(blur_radius)
  )

  rdd %>% gen_raster_image(
    viz_op = viz_op,
    output_location = output_location,
    output_format = output_format,
    overlay = overlay
  )
  if (browse) {
    utils::browseURL(paste0(output_location, ".", tolower(output_format)))
  }

  invisible(viz_op %>% new_viz_op("heatmap"))
}

#' Visualize a Sedona spatial RDD using a scatter plot.
#'
#' Generate a scatter plot of geometrical object(s) within a Sedona spatial RDD.
#'
#' @inheritParams sedona_visualization_routines
#' @param reverse_coords Whether to reverse spatial coordinates in the plot
#'   (default: FALSE).
#'
#' @examples
#' library(sparklyr)
#' library(sparklyr.sedona)
#'
#' sc <- spark_connect(master = "spark://HOST:PORT")
#'
#' if (!inherits(sc, "test_connection")) {
#'   input_location <- system.file(
#'     file.path("extdata", "arealm-small.csv"), package = "sparklyr.sedona"
#'   )
#'   rdd <- sedona_read_dsv_to_typed_rdd(
#'     sc,
#'     location = input_location,
#'     type = "point"
#'   )
#'
#'   sedona_render_scatter_plot(
#'     rdd,
#'     resolution_x = 800,
#'     resolution_y = 600,
#'     output_location = tempfile("arealm-small-"),
#'     output_format = "png",
#'     boundary = c(-91, -84, 30, 35)
#'   )
#' }
#'
#' @family Sedona visualization routines
#' @export
sedona_render_scatter_plot <- function(
                                       rdd,
                                       resolution_x,
                                       resolution_y,
                                       output_location,
                                       output_format = c("png", "gif", "svg"),
                                       boundary = NULL,
                                       color_of_variation = c("red", "green", "blue"),
                                       base_color = c(0, 0, 0),
                                       shade = TRUE,
                                       reverse_coords = FALSE,
                                       overlay = NULL,
                                       browse = interactive()) {
  viz_op <- sedona_render_viz_effect(
    viz_effect_name = "ScatterPlot",
    rdd = rdd,
    resolution_x = resolution_x,
    resolution_y = resolution_y,
    output_location = output_location,
    output_format = output_format,
    boundary = boundary,
    color_of_variation = color_of_variation,
    base_color = base_color,
    shade = shade,
    reverse_coords = reverse_coords,
    overlay = overlay,
    browse = browse
  )

  invisible(viz_op %>% new_viz_op("scatter_plot"))
}

#' Visualize a Sedona spatial RDD using a choropleth map.
#'
#' Generate a choropleth map of a pair RDD assigning integral values to
#' polygons.
#'
#' @inheritParams sedona_visualization_routines
#' @param pair_rdd A pair RDD with Sedona Polygon objects being keys and
#'   java.lang.Long being values.
#' @param reverse_coords Whether to reverse spatial coordinates in the plot
#'   (default: FALSE).
#'
#' @examples
#' library(sparklyr)
#' library(sparklyr.sedona)
#'
#' sc <- spark_connect(master = "spark://HOST:PORT")
#'
#' if (!inherits(sc, "test_connection")) {
#'   pt_input_location <- system.file(
#'     file.path("extdata", "arealm-small.csv"), package = "sparklyr.sedona"
#'   )
#'   pt_rdd <- sedona_read_dsv_to_typed_rdd(
#'     sc,
#'     location = pt_input_location,
#'     type = "point",
#'     first_spatial_col_index = 1
#'   )
#'   polygon_input_location <- system.file(
#'     file.path("extdata", "polygon.json"), package = "sparklyr.sedona"
#'   )
#'   polygon_rdd <- sedona_read_geojson_to_typed_rdd(
#'     sc,
#'     location = polygon_input_location,
#'     type = "polygon"
#'   )
#'   join_result_rdd <- sedona_spatial_join_count_by_key(
#'     pt_rdd,
#'     polygon_rdd,
#'     join_type = "intersect",
#'     partitioner = "quadtree"
#'   )
#'   sedona_render_choropleth_map(
#'     join_result_rdd,
#'     400,
#'     200,
#'     output_location = tempfile("choropleth-map-"),
#'     boundary = c(-86.8, -86.6, 33.4, 33.6),
#'     base_color = c(255, 255, 255)
#'   )
#' }
#'
#' @family Sedona visualization routines
#' @export
sedona_render_choropleth_map <- function(
                                         pair_rdd,
                                         resolution_x,
                                         resolution_y,
                                         output_location,
                                         output_format = c("png", "gif", "svg"),
                                         boundary = NULL,
                                         color_of_variation = c("red", "green", "blue"),
                                         base_color = c(0, 0, 0),
                                         shade = TRUE,
                                         reverse_coords = FALSE,
                                         overlay = NULL,
                                         browse = interactive()) {
  viz_op <- sedona_render_viz_effect(
    viz_effect_name = "ChoroplethMap",
    rdd = pair_rdd,
    resolution_x = resolution_x,
    resolution_y = resolution_y,
    output_location = output_location,
    output_format = output_format,
    boundary = boundary,
    color_of_variation = color_of_variation,
    base_color = base_color,
    shade = shade,
    reverse_coords = reverse_coords,
    overlay = overlay,
    browse = browse
  )

  invisible(viz_op %>% new_viz_op("choropleth_map"))
}

sedona_render_viz_effect <- function(
                                     viz_effect_name,
                                     rdd,
                                     resolution_x,
                                     resolution_y,
                                     output_location,
                                     output_format = c("png", "gif", "svg"),
                                     boundary = NULL,
                                     color_of_variation = c("red", "green", "blue"),
                                     base_color = c(0, 0, 0),
                                     shade = shade,
                                     reverse_coords = FALSE,
                                     overlay = NULL,
                                     browse = interactive()) {
  sc <- spark_connection(rdd$.jobj)
  output_format <- match.arg(output_format)
  color_of_variation <- match.arg(color_of_variation)
  validate_base_color(base_color)

  boundary <- validate_boundary(rdd, boundary)
  viz_op <- invoke_new(
    sc,
    paste0("org.apache.sedona.viz.extension.visualizationEffect.", viz_effect_name),
    as.integer(resolution_x),
    as.integer(resolution_y),
    boundary$.jobj,
    reverse_coords
  )

  rdd %>%
    gen_raster_image(
      viz_op = viz_op,
      output_location = output_location,
      output_format = output_format,
      color_settings = list(
        color_of_variation = color_of_variation,
        base_color = base_color,
        shade = shade
      ),
      overlay = overlay
    )
  if (browse) {
    utils::browseURL(paste0(output_location, ".", tolower(output_format)))
  }

  invisible(viz_op)
}

validate_base_color <- function(base_color) {
  if (!is.numeric(base_color) || length(base_color) != 3) {
    stop(
      "Base color (`base_color`) must be a numeric vector of length 3 ",
      "specifying values for red, green, and blue channels ",
      "(e.g., c(0, 0, 0))."
    )
  }
}

validate_boundary <- function(rdd, boundary) {
  sc <- spark_connection(rdd$.jobj)

  if (is.null(boundary)) {
    minimum_bounding_box(rdd)
  } else if (inherits(boundary, "bounding_box")) {
    boundary
  } else if (is.numeric(boundary)) {
    if (length(boundary) != 4) {
      stop(
        "Boundary specification with numeric vector must consist of ",
        "exactly 4 values: c(min_x, max_x, min_y, max_y)."
      )
    }
    do.call(new_bounding_box, append(list(sc), as.list(boundary)))
  } else {
    stop(
      "Boundary specification must be either NULL, a numeric vector of ",
      "c(min_x, max_x, min_y, max_y) values, or a bounding box object"
    )
  }
}

gen_raster_image <- function(
                             rdd,
                             viz_op,
                             output_location,
                             output_format,
                             color_settings = NULL,
                             overlay = NULL) {
  sc <- spark_connection(rdd$.jobj)

  image_generator <- invoke_new(
    sc,
    "org.apache.sedona.viz.core.ImageGenerator"
  )
  if (!is.null(color_settings)) {
    customize_color_params <- list(viz_op, "CustomizeColor") %>%
      append(as.list(as.integer(unlist(color_settings$base_color)))) %>%
      append(
        list(
          255L, # gamma
          sc$state$enums$awt_color[[color_settings$color_of_variation]],
          color_settings$shade
        )
      )
    do.call(invoke, customize_color_params)
  }
  invoke(viz_op, "Visualize", java_context(sc), rdd$.jobj)

  invoke(
    image_generator,
    "SaveRasterImageAsLocalFile",
    viz_op %>% process_overlay(overlay),
    output_location,
    sc$state$enums$image_types[[output_format]]
  )
}

process_overlay <- function(viz_op, overlay) {
  if (!is.null(overlay)) {
    sc <- spark_connection(viz_op)

    viz_op_img <- invoke(viz_op, "rasterImage")
    overlay_op <- invoke_new(
      sc,
      "org.apache.sedona.viz.core.RasterOverlayOperator",
      viz_op_img
    )
    overlay_img <- invoke(overlay$.jobj, "rasterImage")
    invoke(overlay_op, "JoinImage", overlay_img)
    invoke(overlay_op, "backRasterImage")
  } else {
    invoke(viz_op, "rasterImage")
  }
}

new_viz_op <- function(jobj, type = NULL) {
  structure(list(.jobj = jobj, class = c(type, "viz_op")))
}
